File: atcoder/135/d.py
MOD = 10**9 + 7
memo = None


def rec(S, k, total):
    if k == len(S)//3:
        # print(total)
        fixed_total = total + 5 if k % 2 == 0 else total - 5
        # fixed_total = abs(total) + 5
        memo[k][total % 13] = int((fixed_total) % 13 == 0)
        return int((fixed_total) % 13 == 0)

    if memo[k][total % 13] is not None:
        return memo[k][total % 13]

    s = S[k*3:k*3+3]
    d1 = [str(i) for i in range(10)] if s[0] == '?' else [s[0]]
    d2 = [str(i) for i in range(10)] if s[1] == '?' else [s[1]]
    d3 = [str(i) for i in range(10)] if s[2] == '?' else [s[2]]
    ret = 0

    for i1 in d1:
        for i2 in d2:
            for i3 in d3:
                num = int(i1 + i2 + i3)
                if k % 2 == 0:
                    ret += rec(S, k+1, total + num)
                else:
                    ret += rec(S, k+1, total - num)

    memo[k][total % 13] = ret
    return ret

S = input()

if len(S) % 3 == 1:
    S = '00' + S
elif len(S) % 3 == 2:
    S = '0' + S

L = (len(S) // 3) + 1
memo = [[None for i in range(14)] for j in range(L)]
# print(rec(S, 0, 0) % MOD)
# print(memo)

for k in range(L//3):
    for m in range(13):
        s = S[k*3:k*3+3]
        d1 = [str(i) for i in range(10)] if s[0] == '?' else [s[0]]
        d2 = [str(i) for i in range(10)] if s[1] == '?' else [s[1]]
        d3 = [str(i) for i in range(10)] if s[2] == '?' else [s[2]]

        for i1 in d1:
            for i2 in d2:
                for i3 in d3:
                    num = int(i1 + i2 + i3)
                    if k % 2 == 0:
                        ret += rec(S, k+1, total + num)
                        ret %= MOD
                    else:
                        ret += rec(S, k+1, total - num)
                        ret %= MOD

File: atcoder/135/d.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
	// "math"
	"fmt"
	"math/rand"
	"time"
)

var (
	sc = bufio.NewScanner(os.Stdin)
)

const mod = 1e9 + 7

func main() {
	var s = nextLine()
	var L = len(s)

	dp := make([][]int, L+1)
	for i := 0; i < L+1; i++ {
		dp[i] = make([]int, 13)
	}

	dp[0][0] = 1

	for k := 0; k < L; k++ {
		for m := 0; m < 13; m++ {
			if s[k] == '?' {
				for i := 0; i < 10; i++ {
					dp[k+1][(m * 10 + i) % 13] += dp[k][m]
					dp[k+1][(m * 10 + i) % 13] %= mod
				}
			} else {
				var si = int(s[k] - '0')
				dp[k+1][(m * 10 + si) % 13] += dp[k][m]
				dp[k+1][(m * 10 + si) % 13] %= mod
			}
		}
	}

	fmt.Println(dp[L][5])
}

// Input helpers
func init()  {
	rand.Seed(time.Now().UnixNano())
	buf := make([]byte, 0) //ここのサイズは何でもよさそう
	sc.Buffer(buf, 1000000009)
}

func nextInt() int {
	sc.Scan()
	i, e := strconv.Atoi(sc.Text())
	if e != nil {
		panic(e)
	}
	return i
}

func nextLine() string {
	sc.Scan()
	return sc.Text()
}

func spaceLine() []string {
	sc.Scan()
	return strings.Split(sc.Text()," ")
}

func intLine() []int {
	str := spaceLine()
	buf := make([]int,len(str))
	for i,s:=range str{
		buf[i],_ = strconv.Atoi(s)
	}
	return buf
}
